geode-core/src/main/java/org/apache/geode/distributed/internal/StartupMessage.java

```diff
@@ -173,118 +173,116 @@ public void setEnforceUniqueZone(boolean enforceUniqueZone) {
+          rejectionMessage = String.format(msg, getSender());
+        } else {
+          dm.setEquivalentHosts(this.interfaces);
+        }
       }
-    }
 
-    if (rejectionMessage != null) {
-      logger.warn(rejectionMessage);
-    }
+      if (rejectionMessage != null) {
+        logger.warn(rejectionMessage);
+      }
+
+      if (rejectionMessage == null) { // change state only if there's no rejectionMessage yet
+        dm.setRedundancyZone(getSender(), this.redundancyZone);
+        dm.setEnforceUniqueZone(this.enforceUniqueZone);
+
+        if (this.hostedLocatorsAll != null) {
+          // boolean isSharedConfigurationEnabled = false;
+          // if (this.hostedLocatorsWithSharedConfiguration != null) {
+          // isSharedConfigurationEnabled = true;
+          // }
+          dm.addHostedLocators(getSender(), this.hostedLocatorsAll,
+              this.isSharedConfigurationEnabled);
+        }
+      }
 
-    if (rejectionMessage == null) { // change state only if there's no rejectionMessage yet
-      dm.setRedundancyZone(getSender(), this.redundancyZone);
-      dm.setEnforceUniqueZone(this.enforceUniqueZone);
-
-      if (this.hostedLocatorsAll != null) {
-        // boolean isSharedConfigurationEnabled = false;
-        // if (this.hostedLocatorsWithSharedConfiguration != null) {
-        // isSharedConfigurationEnabled = true;
-        // }
-        dm.addHostedLocators(getSender(), this.hostedLocatorsAll,
-            this.isSharedConfigurationEnabled);
+      StartupResponseMessage m =
+          new StartupResponseWithVersionMessage(dm, replyProcessorId, getSender(), rejectionMessage,
+              isAdminDM);
+      if (logger.isDebugEnabled()) {
+        logger.debug("Received StartupMessage from a member with version: {}, my version is:{}",
+            this.version, GemFireVersion.getGemFireVersion());
+      }
+      dm.putOutgoing(m);
+      if (rejectionMessage != null) {
+        dm.getMembershipManager().startupMessageFailed(getSender(), rejectionMessage);
       }
-    }
 
+      // We need to discard this member if they aren't a peer.
+      if (rejectionMessage != null) {
+        dm.handleManagerDeparture(getSender(), false, rejectionMessage);
+      }
 
-    StartupResponseMessage m = null;
-    // Commenting out. See Bruces note in the StartupMessageData constructor.
-    // Comparisons should use the functionality described in SerializationVersions
-    // if (GemFireVersion.compareVersions(theirVersion,"6.6.2") >= 0) {
-    m = new StartupResponseWithVersionMessage(dm, replyProcessorId, getSender(), rejectionMessage,
-        isAdminDM);
-    // } else {
-    // m = new StartupResponseMessage(dm, replyProcessorId, getSender(), rejectionMessage,
-    // isAdminDM);
-    // }
-    if (logger.isDebugEnabled()) {
-      logger.debug("Received StartupMessage from a member with version: {}, my version is:{}",
-          theirVersion, myVersion);
+    } catch (RuntimeException e) {
```

Comment:
I expected you'd have a finally block that ensures a response is sent unless we're shutting down, but your finally block is empty.   Would you address that?  I can work with you on it if you like.